pkg/kubectl/cmd/edit.go

```diff
@@ -0,0 +1,419 @@
+	file      string
+
+	delta   *jsonmerge.Delta
+	version string
+}
+
+func (r *editResults) AddError(err error, info *resource.Info) string {
+	switch {
+	case errors.IsInvalid(err):
+		r.edit = append(r.edit, info)
+		reason := editReason{
+			head: fmt.Sprintf("%s %s was not valid", info.Mapping.Kind, info.Name),
+		}
+		if err, ok := err.(kclient.APIStatus); ok {
+			if details := err.Status().Details; details != nil {
+				for _, cause := range details.Causes {
+					reason.other = append(reason.other, cause.Message)
+				}
+			}
+		}
+		r.header.reasons = append(r.header.reasons, reason)
+		return fmt.Sprintf("Error: the %s %s is invalid", info.Mapping.Kind, info.Name)
+	case errors.IsNotFound(err):
+		r.notfound++
+		return fmt.Sprintf("Error: the %s %s has been deleted on the server", info.Mapping.Kind, info.Name)
+
+	case errors.IsConflict(err):
+		if r.delta != nil {
+			v1 := info.ResourceVersion
+			if perr := applyPatch(r.delta, info, r.version); perr != nil {
+				// the error was related to the patching process
+				if nerr, ok := perr.(patchError); ok {
+					r.conflict++
+					if jsonmerge.IsPreconditionFailed(nerr.error) {
+						return fmt.Sprintf("Error: the API version of the provided object cannot be changed")
+					}
+					// the patch is in conflict, report to user and exit
+					if jsonmerge.IsConflicting(nerr.error) {
+						// TODO: read message
+						return fmt.Sprintf("Error: a conflicting change was made to the %s %s on the server", info.Mapping.Kind, info.Name)
+					}
+					glog.V(4).Infof("Attempted to patch the resource, but failed: %v", perr)
+					return fmt.Sprintf("Error: %v", err)
+				}
+				// try processing this server error and unset delta so we don't recurse
+				r.delta = nil
+				return r.AddError(err, info)
+			}
+			return fmt.Sprintf("Applied your changes to %s from version %s onto %s", info.Name, v1, info.ResourceVersion)
+		}
+		// no delta was available
+		r.conflict++
+		return fmt.Sprintf("Error: %v", err)
+	default:
+		r.retryable++
+		return fmt.Sprintf("Error: the %s %s could not be updated: %v", info.Mapping.Kind, info.Name, err)
+	}
+}
+
+type patchError struct {
+	error
+}
+
+// applyPatch reads the latest version of the object, writes it to version, then attempts to merge
+// the changes onto it without conflict. If a conflict occurs jsonmerge.IsConflicting(err) is
+// true. The info object is mutated
+func applyPatch(delta *jsonmerge.Delta, info *resource.Info, version string) error {
+	if err := info.Get(); err != nil {
+		return patchError{err}
+	}
+	obj, err := resource.AsVersionedObject([]*resource.Info{info}, false, version)
+	if err != nil {
+		return patchError{err}
+	}
+	data, err := info.Mapping.Codec.Encode(obj)
+	if err != nil {
+		return patchError{err}
+	}
+	merged, err := delta.Apply(data)
+	if err != nil {
+		return patchError{err}
+	}
+	updated, err := resource.NewHelper(info.Client, info.Mapping).Update(info.Namespace, info.Name, false, merged)
```

Comment:
@caesarxuchao as we discussed the two approaches (or three) should be essentially the same in practice. Let's choose the one that we can do in the least lines of code, and that makes our client the skinniest.